`autoretry_for` + `retry_backoff` mutate a single `retry_kwargs` dict shared across all executions of a task, corrupting backoff under concurrent execution

# Checklist

* [x] I have verified that the issue exists against the `main` branch of Celery.
* [ ] This has already been asked to the [[discussions forum](https://github.com/celery/celery/discussions)](https://github.com/celery/celery/discussions) first.
* [ ] I have read the relevant section in the
  [[contribution guide](https://docs.celeryq.dev/en/main/contributing.html#other-bugs)](https://docs.celeryq.dev/en/main/contributing.html#other-bugs)
  on reporting bugs.
* [ ] I have checked the [[issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
* [ ] I have checked the [[pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
* [x] I have checked the [[commit log](https://github.com/celery/celery/commits/main)](https://github.com/celery/celery/commits/main)
  to find out if the bug was already fixed in the main branch.
* [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).
* [ ] I have tried to reproduce the issue with [[pytest-celery](https://docs.celeryq.dev/projects/pytest-celery/en/latest/userguide/celery-bug-report.html)](https://docs.celeryq.dev/projects/pytest-celery/en/latest/userguide/celery-bug-report.html) and added the reproduction script below.

## Mandatory Debugging Information

* [x] I have included the output of `celery -A proj report` in the issue.
  (if you are not able to do this, then at least specify the Celery
  version affected).
* [x] I have verified that the issue exists against the `main` branch of Celery.
* [x] I have included the contents of `pip freeze` in the issue.
* [x] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

* [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
* [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
* [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
* [ ] I have tried reproducing the issue on more than one operating system.
* [ ] I have tried reproducing the issue on more than one workers pool.
* [x] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
* [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

## Related Issues and Possible Duplicates

#### Related Issues

* None found

#### Possible Duplicates

* None found

## Environment & Settings

**Celery version**: 5.6.2

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```text
software -> celery:5.6.2 (immunity) kombu:5.6.2 py:3.12.3
            billiard:4.2.4
platform -> system:Linux arch:64bit
result_backend: disabled
```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

* **Minimal Python Version**: N/A or Unknown
* **Minimal Celery Version**: 5.6.2 (bug is present in `main` as well; the affected code has not changed since at least the 5.6.x line)
* **Minimal Kombu Version**: N/A — not broker-related, bug is purely in `celery.app.autoretry`
* **Minimal Broker Version**: N/A or Unknown
* **Minimal Result Backend Version**: N/A or Unknown
* **Minimal OS and/or Kernel Version**: N/A or Unknown
* **Minimal Broker Client Version**: N/A or Unknown
* **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```text
amqp==5.3.1
billiard==4.2.4
celery==5.6.2
click==8.3.1
click-didyoumean==0.3.1
click-plugins==1.1.1.2
click-repl==0.3.0
kombu==5.6.2
vine==5.1.0
```

</p>
</details>

### Other Dependencies

<details>
<p>

No broker or result backend is needed to reproduce this — the bug is entirely
in-process, in `celery.app.autoretry.add_autoretry_behaviour`, and can be
demonstrated by calling a task's `run()` directly.

</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
"""
add_autoretry_behaviour() builds the `retry_kwargs` dict ONCE, when a task
class is registered (celery/app/base.py calls it exactly once per task
class, never per invocation). That single dict is captured by the
generated run() closure and reused for the entire lifetime of the worker
process -- but run() mutates it in place on every retry:

    retry_kwargs['countdown'] = get_exponential_backoff_interval(...)
    ...
    ret = task.retry(exc=exc, **retry_kwargs)

This script proves the shared dict gets permanently mutated by a task
execution, which is not supposed to happen, and which is unsafe under any
concurrent execution of the same task (eventlet/gevent pools, or any other
cooperative/threaded execution of the same task within one process).
"""
from unittest.mock import patch

from celery import Celery
from celery.exceptions import Retry


app = Celery(set_as_current=False)


@app.task(
    bind=True,
    shared=False,
    autoretry_for=(ZeroDivisionError,),
    retry_backoff=True,
    retry_jitter=False,
    max_retries=3,
)
def task(self_, x, y):
    return x / y


# Locate the `retry_kwargs` dict captured by the run() closure at
# registration time -- this is the exact object that is shared across
# every future execution of this task in the same worker process.
free_vars = task.run.__code__.co_freevars
shared_retry_kwargs = task.run.__closure__[
    free_vars.index("retry_kwargs")
].cell_contents

print("shared retry_kwargs BEFORE any retries:", shared_retry_kwargs)

assert shared_retry_kwargs == {}


with patch.object(task, "retry", side_effect=Retry):
    for retries in range(3):
        task.push_request(retries=retries)

        try:
            task.run(1, 0)
        except Retry:
            pass
        finally:
            task.pop_request()


print("shared retry_kwargs AFTER 3 retries:  ", shared_retry_kwargs)

assert shared_retry_kwargs == {}, (
    "BUG: shared retry_kwargs dict was mutated!"
)
```

</p>
</details>

# Expected Behavior

The `retry_kwargs` dict captured by the `run()` closure at task-registration
time should never be mutated.

Running the script above should print:

```text
shared retry_kwargs BEFORE any retries: {}
shared retry_kwargs AFTER 3 retries:   {}
```

and complete without raising `AssertionError`.

More importantly, two concurrent executions of the same task (e.g. under an
eventlet/gevent pool) should each retry using **their own** backoff
countdown, computed from their own `task.request.retries`, and should never
be able to see or use a countdown/`max_retries` value written by a
*different, concurrently-running* execution of the same task.

# Actual Behavior

Running the script above prints:

```text
shared retry_kwargs BEFORE any retries: {}
shared retry_kwargs AFTER 3 retries:    {'countdown': 4}

Traceback (most recent call last):
  ...
AssertionError: BUG: shared retry_kwargs dict was mutated!
```

The single `retry_kwargs` dict shared by every execution of the task
permanently picks up a `'countdown'` key (and, if `override_max_retries` is
used, a `'max_retries'` key) from whichever execution most recently retried.

Because it is the *same dict object* across all executions of the task for
the whole lifetime of the process, this is unsafe under concurrent execution
of the same task (eventlet/gevent pools, or any other cooperative/threaded
execution of the same task in one process): one execution's write of
`countdown`/`max_retries` can land in the window between another execution's
write and its use of that dict in `task.retry(**retry_kwargs)`, so that other
execution ends up retrying with the **wrong backoff delay** (or the wrong
`max_retries`).

This silently defeats `retry_backoff`/`retry_jitter`, whose purpose is
specifically to prevent a thundering herd of retries against a struggling
downstream service.

It also means that if a user sets:

```python
retry_kwargs = {"max_retries": 5}
```

as a task **class attribute**, this code mutates that exact class-attribute
dict in place, permanently injecting a `'countdown'` key that was never
there.

**Affected file:** `celery/app/autoretry.py` (the `except autoretry_for as
exc:` branch inside the generated `run()` closure in
`add_autoretry_behaviour`).

**Root cause:** `retry_kwargs` is built once per task class at
registration time (`celery/app/base.py`, `Celery.task()` and
`celery/app/registry.py`, `register_task()` each call
`add_autoretry_behaviour(task, **options)` exactly once) and then mutated
in place on every retry instead of being copied per call.

A possible fix seems to be working on a per-invocation copy of the dictionary inside the `except` block instead of mutating the closed-over `retry_kwargs`.

If this behavior is confirmed as a bug and that approach looks appropriate, I'd be happy to work on the fix and add a regression test.

--- a/celery/app/autoretry.py
+++ b/celery/app/autoretry.py
@@ -45,8 +45,9 @@ def add_autoretry_behaviour(task, **options):
             except dont_autoretry_for:
                 raise
             except autoretry_for as exc:
+                retry_kwargs_for_attempt = retry_kwargs.copy()
                 if retry_backoff:
-                    retry_kwargs['countdown'] = \
+                    retry_kwargs_for_attempt['countdown'] = \
                         get_exponential_backoff_interval(
                             factor=int(max(1.0, retry_backoff)),
                             retries=task.request.retries,
@@ -54,10 +55,9 @@ def add_autoretry_behaviour(task, **options):
                             full_jitter=retry_jitter)
                 # Override max_retries
                 if hasattr(task, 'override_max_retries'):
-                    retry_kwargs['max_retries'] = getattr(task,
-                                                          'override_max_retries',
-                                                          task.max_retries)
-                ret = task.retry(exc=exc, **retry_kwargs)
+                    retry_kwargs_for_attempt['max_retries'] = getattr(
+                        task, 'override_max_retries', task.max_retries)
+                ret = task.retry(exc=exc, **retry_kwargs_for_attempt)
                 # Stop propagation
                 if hasattr(task, 'override_max_retries'):
                     delattr(task, 'override_max_retries')

--- a/t/unit/tasks/test_tasks.py
+++ b/t/unit/tasks/test_tasks.py
@@ -1785,6 +1785,41 @@ class test_apply_task(TasksCase):
             assert data['parent_id'] == 'parent-id-123'
             assert data['root_id'] == data['task_id']
 
+    def test_autoretry_shared_retry_kwargs_not_mutated_by_max_retries_override(self):
+        """Verify that task.retry(max_retries=...) setting override_max_retries
+        does not mutate the shared retry_kwargs dict captured at task
+        registration time (regression test for celery#10456)."""
+        @self.app.task(bind=True, shared=False,
+                       autoretry_for=(ZeroDivisionError,),
+                       retry_kwargs={'max_retries': 5},
+                       retry_backoff=False, retry_jitter=False)
+        def task(self_, x, y):
+            self_.iterations += 1
+            return x / y
+        task.iterations = 0
+
+        free_vars = task.run.__code__.co_freevars
+        shared_retry_kwargs = task.run.__closure__[
+            free_vars.index("retry_kwargs")
+        ].cell_contents
+        assert shared_retry_kwargs == {'max_retries': 5}
+
+        task.override_max_retries = 2
+
+        with patch.object(task, "retry", side_effect=Retry):
+            task.push_request(retries=0)
+            try:
+                task.run(1, 0)
+            except Retry:
+                pass
+            finally:
+                task.pop_request()
+
+        assert shared_retry_kwargs == {'max_retries': 5}, (
+            "BUG: override_max_retries leaked into shared retry_kwargs! "
+            f"Got {shared_retry_kwargs}"
+        )
+
 
 class test_apply_async(TasksCase):
     def common_send_task_arguments(self):

--- a/t/unit/tasks/test_tasks.py
+++ b/t/unit/tasks/test_tasks.py
@@ -833,6 +833,51 @@ class test_task_retries(TasksCase):
         ]
         assert retry_call_countdowns == expected_countdowns
 
+    def test_autoretry_does_not_mutate_shared_base_class_retry_kwargs(self):
+        """
+        Test that retry_kwargs defined as a class attribute on a shared base Task
+        is not mutated when a task retries. This covers the getattr(task, 'retry_kwargs', {})
+        branch in add_autoretry_behaviour, which is worse than the per-task case since
+        the same dict object is shared across all task classes using that base.
+        """
+        class BaseTaskWithRetry(Task):
+            autoretry_for = (ZeroDivisionError,)
+            retry_backoff = True
+            retry_jitter = False
+            retry_kwargs = {'max_retries': 5}
+            _app = self.app
+
+        @self.app.task(bind=True, base=BaseTaskWithRetry, shared=False)
+        def task_a(self_, x, y):
+            return x / y
+
+        @self.app.task(bind=True, base=BaseTaskWithRetry, shared=False)
+        def task_b(self_, x, y):
+            return x / y
+
+        # Verify that all three share the exact same dict object
+        assert task_a.retry_kwargs is task_b.retry_kwargs is BaseTaskWithRetry.retry_kwargs
+
+        # Run task_a to exhaustion, which would mutate the shared dict if bug exists
+        with patch.object(task_a, 'retry', wraps=task_a.retry) as fake_retry:
+            task_a.apply((1, 0))
+
+        # Base class retry_kwargs should remain unchanged (no 'countdown' key leaked in)
+        assert BaseTaskWithRetry.retry_kwargs == {'max_retries': 5}
+
+        # Verify task_a got correct countdown values
+        assert [call_[1]['countdown'] for call_ in fake_retry.call_args_list] == [
+            1, 2, 4, 8, 16, 32
+        ]
+
+        # Run task_b once and verify it uses its own first-attempt backoff,
+        # not something inherited from task_a's last retry
+        with patch.object(task_b, 'retry', wraps=task_b.retry) as fake_retry_b:
+            task_b.apply((1, 0))
+
+        # First retry should have countdown=1 (first attempt backoff)
+        assert fake_retry_b.call_args_list[0][1]['countdown'] == 1
+
 
 class test_canvas_utils(TasksCase):
 

